### PR TITLE
Feat/dex router

### PR DESCRIPTION
**Pull request by lubkoll**
Current state:
- swaps work
- cw20 removed
- changed computation of best path to use single query
- on-chain computation of optimal swap route
- Rename to dex-router-osmosis

### CODE REVIEW

File: smart-contracts/Cargo.toml [hunk 1]
```diff
@@ -32,6 +33,8 @@ cw20-base = {version = "1.1.2", features = ["library"]}
 cw20-staking = "0.11.1"
 apollo-cw-asset = "0.1.2"
 cw-dex = {git = "https://github.com/quasar-finance/cw-dex", branch = "feat/deprecate-osmo-gamm"}
+mars-owner = "2.0.0"
```
Comment (magiodev) [inline smart-contracts/Cargo.toml:36]: I dont know if we should avoid add this dependency only to use the NotOwner error. Can we replace it for a local Unauthorized {} enum variant on the error.rs?

File: smart-contracts/contracts/dex-router-osmosis/src/msg.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        offer_denom: String,
+        ask_denom: String,
+        path: Vec<u64>,
+        bidirectional: bool,
+    },
+    RemovePath {
+        offer_denom: String,
+        ask_denom: String,
+        path: Vec<u64>,
+        bidirectional: bool,
```
Comment (magiodev) [inline smart-contracts/contracts/dex-router-osmosis/src/msg.rs:26]: Can this field be removed? I think we only need the index selectors here during the remove.

File: smart-contracts/contracts/dex-router-osmosis/src/state.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+use mars_owner::Owner;
+use osmosis_std::types::osmosis::poolmanager::v1beta1::SwapAmountInRoute;
+
+#[cw_serde]
+pub struct RecipientInfo {
+    pub address: Addr,
+    pub denom: String,
+}
+
+pub const PATHS: Map<(String, String), Vec<Vec<SwapAmountInRoute>>> = Map::new("paths");
```
Comment (magiodev) [inline smart-contracts/contracts/dex-router-osmosis/src/state.rs:13]: Can we comment at line 12 what the index String, String represents?

**Review by magiodev — CHANGES_REQUESTED** [review]
Mostly looks good! Minor changes requested.

File: smart-contracts/contracts/dex-router-osmosis/src/state.rs (hunk 3 above)
Comment (lubkoll) [reply smart-contracts/contracts/dex-router-osmosis/src/state.rs:14]: done

File: smart-contracts/contracts/dex-router-osmosis/src/msg.rs (hunk 2 above)
Comment (lubkoll) [reply smart-contracts/contracts/dex-router-osmosis/src/msg.rs:26]: Yes. I have it here for symmetry. Both in SetPath and RemovePath we could leave it out. As it is in SetPath, the interface looks a bit more consistent to me with having it here too. I can remove it if you prefer to, else I will keep it.

File: smart-contracts/Cargo.toml (hunk 1 above)
Comment (lubkoll) [reply smart-contracts/Cargo.toml:36]: As these checks for the sender are the same in all contracts it is quite nice to reuse existing solutions for that. The mars-owner is just the new version of the Admin in cw-controllers. It has a better implementation and -- though not used here -- a quite safe mechanism for transfer of ownership.

File: smart-contracts/contracts/dex-router-osmosis/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Dex router for osmosis
+
+Responsible for the on-chain computation of multi-hop swap routes.
+Currently only supports pools with 2 tokens.
+
+TODO:
+ * Fix queries for best path
+ * Add checks and unit-tests
+ * Add support for (optional) on-chain detection of the optimal swap path
```
Comment (arhamchordia) [inline smart-contracts/contracts/dex-router-osmosis/README.md:9]: Can add a todo for test-tube tests.

File: smart-contracts/Cargo.toml (hunk 1 above)
Comment (magiodev) [reply smart-contracts/Cargo.toml:36]: I was going to suggest to use cw-controllers but ok, sounds good!

File: smart-contracts/contracts/dex-router-osmosis/README.md (hunk 4 above)
Comment (lubkoll) [reply smart-contracts/contracts/dex-router-osmosis/README.md]: Thx, actually that list wasn't updated properly by me. Fixed it.
Now, the only item is "possibly test-tube tests". Wrote possibly as I have some concerns that that might be more work than we get in return from it.

File: smart-contracts/contracts/dex-router-osmosis/src/msg.rs (hunk 2 above)
Comment (magiodev) [reply smart-contracts/contracts/dex-router-osmosis/src/msg.rs:25]: If it is useful we can keep it. If not we should remove

File: smart-contracts/contracts/dex-router-osmosis/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    _msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    OWNER.initialize(
```
Comment (ajansari95) [inline smart-contracts/contracts/dex-router-osmosis/src/contract.rs:34]: here `OWNER` -(admin) is set on the sender address, which cant be updated afterwards. We should have a `updatOwner` or `updateAfmin` endpoint. @lubkoll @magiodev ?

Comment (lubkoll) [reply smart-contracts/contracts/dex-router-osmosis/src/contract.rs:34]: done

**Review by ajansari95 — DISMISSED** [review]
lgtm

**Review by magiodev — DISMISSED** [review]
looks good!

**Review by magiodev — APPROVED** [review]

**Review by ajansari95 — APPROVED** [review]
lgtm

**Review by arhamchordia — APPROVED** [review]
LGTM

**Pull request closed without merging**